### PR TITLE
Use miekg/dns for SRV lookup

### PR DESCRIPTION
**Pull request by codesome**
Fixes https://github.com/cortexproject/cortex/issues/1200

EDIT (by @gouthamve): This uses miekg/dns library to do the SRV lookups. We are using this in the grpc codebase as there is a A record fallback when SRV lookups fail.

### CODE REVIEW

**Comment by tomwilkie** [comment]
Looks good so far; lint needs fixing, and there was some mention of gRPC also using LookupSRV for querier -> frontend comms.

**Comment by codesome** [comment]
Yes, it was [this](https://github.com/cortexproject/cortex/issues/1200#issuecomment-457941931) comment by @bboreham. Need to fix that too. Will do when I get free from academic stuff.

**Comment by tomwilkie** [comment]
I opened https://github.com/grpc/grpc-go/pull/2634 to help with the gRPC side.

**Comment by codesome** [comment]
Thanks @tomwilkie!

File: pkg/util/dns.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	msg := &dns.Msg{}
+	name = "_" + service + "._" + proto + "." + name
+	msg.SetQuestion(dns.Fqdn(name), dns.TypeSRV)
+
+	var result []*dns.SRV
+	dnsResolved := false
+	for _, serverAddr := range dnsServers {
+		resMsg, err := dns.Exchange(msg, serverAddr)
+		if err != nil {
+			continue
```
Comment (gouthamve) [inline pkg/util/dns.go]: Please log the error here. Deploying this, gives this error: `dial udp: address 10.47.240.10: missing port in address`

**Comment by bboreham** [comment]
Note the gRPC PR was rejected

**Comment by gouthamve** [comment]
Hmm, as far as I see, the grpc resolver uses `grpclb` port name to resolve the services, https://github.com/grpc/grpc-go/blob/master/naming/dns_resolver.go#L216-L220

If that fails, it falls back to simple A record lookup: https://github.com/grpc/grpc-go/blob/master/naming/dns_resolver.go#L261-L267

I found that in the clusters we run, we don't expose a port-named `grpclb` and the resolver falls back to A records. Given that it works nevertheless, I wonder if it's worth the pain to copy the `dns_resolver.go` file from `grpc-go`. Note that this doesn't exist in the latest kubernetes versions and only is a problem when running cortex with `go 1.11` with an old kubernetes version. Even then nothing will break.

**Comment by bboreham** [comment]
I'm happy with A record lookup.
Can someone update #1200 with exactly what works and doesn't work?
(e.g. what's "an old kubernetes version")

**Comment by gouthamve** [comment]
A: We don't need to do it if we're okay asking people to upgrade kube-dns
B: The TCP-UDP change has been reverted in the last commit by me. The current state of the PR could be described as:

Use miekg/dns instead of native DNS library to be compatible with kube-dns `<1.14.11` and golang `>=1.11`. This is what Prometheus does for DNS-SD

C: The problem is that `go1.11`'s net library cannot read SRV records sent by kube-dns `<1.14.11`

**Comment by bboreham** [comment]
Still feels like a bad idea to me. 
Note I have two questions outstanding on the go modules PR #1232 for 25 days.

**Comment by tomwilkie** [comment]
> Still feels like a bad idea to me.

What feels bad about it?  Can I infer from your comments you'd prefer to upgrade go without this, effectively making Cortex not work on kube-dns <1.14.11?

> Note I have two questions outstanding on the go modules PR #1232 for 25 days.

Thats a different PR; lets address that there, it shouldn't block this.

**Comment by bboreham** [comment]
> What feels bad about it?

It's adding 70 lines of code in a client to work around an issue in a server which was fixed months ago.

Are there reports from people who use or would like to use Cortex and cannot upgrade kube-dns?

Also I think there should be comments in the code
 - why we are doing this
 - why do it in some places and not others
 - why you shouldn't call this code very often
 - why it's ok to have a relatively inefficient implementation

**Comment by gouthamve** [comment]
Hmm, I'm going to hold off on this PR until someone actually raises an issue and proceed with Go 1.12 upgrade. We might just be optimising for something that has been fixed already as you point out.

**Comment by tomwilkie** [comment]
@khaines on the call yesterday we decided we probably didn't need this, but we just want to check you've got an update kubedns?

**Comment by tomwilkie** [comment]
No word from @khaines. I'll close, we can reopen if needs be.

**Pull request closed without merging**